File: Dizzo-Studio-Frontend/app/lib/design/dial.ts
// A clock face's numerals and minute marks. Everything is placed along
// rays from the centre, 6° apart (every minute, 12 at the top, clockwise):
// on a round face they sit on a circle (an ellipse in a non-square box),
// on a rectangular one where each ray meets the rounded rectangle, the way
// square wall clocks are laid out. Sizes follow the face, so the marks are
// right for any clock; the numerals take the customer's font and size.
import type { DialSource, Layer } from '~/lib/design/document';
import { areaMethod, zoneBox } from '~/lib/design/document';
import { isPlacedModelAnchor } from '~/types/catalog';
import type { PrintArea } from '~/types/catalog';

const ROMAN = ['XII', 'I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX', 'X', 'XI'];
export const DIAL_PREFIX = 'dial-';

/** The face edge along the ray at `angle` (radians, 0 = up, clockwise): its
 * distance from the centre, for a face of w × h (mm or px alike). */
function edgeDistance(dial: Pick<DialSource, 'face' | 'corner_radius_mm'>, w: number, h: number, angle: number, cornerPx: number): number {
  const dx = Math.sin(angle);
  const dy = -Math.cos(angle);
  const hw = w / 2;
  const hh = h / 2;
  if (dial.face === 'round') return 1 / Math.sqrt((dx / hw) ** 2 + (dy / hh) ** 2);
  const r = Math.min(cornerPx, hw, hh);
  // Inside a rounded rectangle (signed distance ≤ 0), found by bisection.
  const inside = (t: number) => {
    const qx = Math.abs(dx * t) - (hw - r);
    const qy = Math.abs(dy * t) - (hh - r);
    return Math.hypot(Math.max(qx, 0), Math.max(qy, 0)) + Math.min(Math.max(qx, qy), 0) - r <= 0;
  };
  let lo = 0;
  let hi = Math.hypot(hw, hh);
  for (let i = 0; i < 32; i++) {
    const mid = (lo + hi) / 2;
    if (inside(mid)) lo = mid;
    else hi = mid;
  }
  return lo;
}

/** Draws the dial into the box (-w/2, -h/2)–(w/2, h/2) of the current transform. */
export function drawDial(
  ctx: CanvasRenderingContext2D, dial: DialSource, w: number, h: number, s: number, color: string, font: (px: number) => string,
) {
  const d = Math.min(w, h);
  const margin = d * 0.022;
  const hourTick = dial.ticks ? d * 0.055 : 0;
  const minuteTick = d * 0.026;
  const corner = dial.corner_radius_mm * s;
  ctx.save();
  ctx.fillStyle = color;
  ctx.strokeStyle = color;
  ctx.lineCap = 'butt';
  if (dial.ticks) {
    for (let i = 0; i < 60; i++) {
      const angle = (i * Math.PI) / 30;
      const hour = i % 5 === 0;
      const edge = edgeDistance(dial, w, h, angle, corner) - margin;
      const len = hour ? hourTick : minuteTick;
      const dx = Math.sin(angle);
      const dy = -Math.cos(angle);
      ctx.lineWidth = hour ? d * 0.011 : d * 0.0045;
      ctx.beginPath();
      ctx.moveTo(dx * edge, dy * edge);
      ctx.lineTo(dx * (edge - len), dy * (edge - len));
      ctx.stroke();
    }
  }
  if (dial.numerals !== 'none') {
    const px = dial.size_mm * s;
    ctx.font = font(px);
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    for (let i = 0; i < 12; i++) {
      const angle = (i * Math.PI) / 6;
      const label = dial.numerals === 'roman' ? ROMAN[i]! : String(i === 0 ? 12 : i);
      const m = ctx.measureText(label);
      // Far enough in that the whole numeral clears the marks, whichever way the ray runs.
      const dx = Math.sin(angle);
      const dy = -Math.cos(angle);
      const half = Math.abs(dx) * (m.width / 2) + Math.abs(dy) * (px * 0.42);
      const r = edgeDistance(dial, w, h, angle, corner) - margin - hourTick - d * 0.02 - half;
      ctx.fillText(label, dx * r, dy * r + px * 0.04);
    }
  }
  ctx.restore();
}

// Measuring the numerals for a hit test (the same text the dial draws).
let measureCtx: CanvasRenderingContext2D | null | undefined;
function measurer(): CanvasRenderingContext2D | null {
  if (measureCtx === undefined) measureCtx = import.meta.client ? document.createElement('canvas').getContext('2d') : null;
  return measureCtx;
}

/** Whether a press lands on what the dial actually draws — a numeral or a
 * mark — and not merely inside its face: the layer covers the whole face,
 * so without this the dial would be selected anywhere on the clock. `p` is
 * in the layer's own frame (mm from the centre, y down) and every mark is
 * widened by `slack` mm so a finger still catches it. */
export function dialHit(
  dial: DialSource, w: number, h: number, p: { x: number; y: number }, font: (px: number) => string, slack = 2,
): boolean {
  const d = Math.min(w, h);
  const margin = d * 0.022;
  const hourTick = dial.ticks ? d * 0.055 : 0;
  const minuteTick = d * 0.026;
  const corner = dial.corner_radius_mm;
  const r = Math.hypot(p.x, p.y);
  if (dial.ticks) {
    // The mark nearest the press runs along its own ray (every 6°).
    const angle = Math.atan2(p.x, -p.y);
    const minute = Math.round((angle * 30) / Math.PI);
    const ray = (minute * Math.PI) / 30;
    const across = Math.abs(r * Math.sin(angle - ray));
    const edge = edgeDistance(dial, w, h, ray, corner) - margin;
    const len = minute % 5 === 0 ? hourTick : minuteTick;
    if (across <= slack && r <= edge + slack && r >= edge - len - slack) return true;
  }
  if (dial.numerals !== 'none') {
    const px = dial.size_mm;
    const ctx = measurer();
    if (ctx) ctx.font = font(px);
    for (let i = 0; i < 12; i++) {
      const angle = (i * Math.PI) / 6;
      const label = dial.numerals === 'roman' ? ROMAN[i]! : String(i === 0 ? 12 : i);
      const width = ctx ? ctx.measureText(label).width : px * 0.62 * label.length;
      const dx = Math.sin(angle);
      const dy = -Math.cos(angle);
      const half = Math.abs(dx) * (width / 2) + Math.abs(dy) * (px * 0.42);
      const at = edgeDistance(dial, w, h, angle, corner) - margin - hourTick - d * 0.02 - half;
      if (Math.abs(p.x - dx * at) <= width / 2 + slack && Math.abs(p.y - (dy * at + px * 0.04)) <= px * 0.6 + slack) return true;
    }
  }
  return false;
}

/** The face an area's clock numerals follow (from its model anchor). */
export function faceOf(area: PrintArea): Pick<DialSource, 'face' | 'corner_radius_mm'> {
  const a = isPlacedModelAnchor(area.anchor) ? area.anchor : null;
  return { face: a?.round ? 'round' : 'rect', corner_radius_mm: a?.corner_radius_mm ? Number(a.corner_radius_mm) : 0 };
}

export const isDialArea = (area: PrintArea) => isPlacedModelAnchor(area.anchor) && area.anchor.dial === true;

/** A new clock face's numerals: covering the colour-print zone, locked by the Studio. */
export function newDialLayer(area: PrintArea, color: string): Layer | null {
  const m = areaMethod(area, 'uv') ?? area.methods[0];
  if (!m) return null;
  const z = zoneBox(m);
  const w = z.x1 - z.x0;
  const h = z.y1 - z.y0;
  return {
    id: `${DIAL_PREFIX}${Date.now().toString(36)}`, area: area.key, method: m.method, kind: 'dial', locked: 'system',
    x_mm: (z.x0 + z.x1) / 2, y_mm: (z.y0 + z.y1) / 2, w_mm: w, h_mm: h, rotation: 0,
    dial: {
      font: 'Montserrat', size_mm: Math.max(6, Math.round(Math.min(w, h) * 0.075)), color: m.colors_allowed ? color : '#000000',
      bold: true, italic: false, numerals: 'arabic', ticks: true, ...faceOf(area),
    },
  };
}

File: Dizzo-Studio-Mobile/lib/features/editor/domain/dial.dart
import 'dart:math' as math;

import 'design_document.dart';
import 'print_area.dart';

// A clock face's numerals and minute marks (the web's `app/lib/design/dial.ts`).
// Everything sits on rays from the centre, 6° apart, 12 at the top.

const romanNumerals = ['XII', 'I', 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX', 'X', 'XI'];

String dialLabel(DialSource dial, int hour) =>
    dial.numerals == 'roman' ? romanNumerals[hour] : '${hour == 0 ? 12 : hour}';

/// The face edge along the ray at [angle] (radians, 0 = up, clockwise).
double dialEdgeDistance(String face, double w, double h, double angle, double corner) {
  final dx = math.sin(angle);
  final dy = -math.cos(angle);
  final hw = w / 2;
  final hh = h / 2;
  if (face == 'round') return 1 / math.sqrt(math.pow(dx / hw, 2) + math.pow(dy / hh, 2));
  final r = math.min(corner, math.min(hw, hh));
  bool within(double t) {
    final qx = (dx * t).abs() - (hw - r);
    final qy = (dy * t).abs() - (hh - r);
    final outside = math.sqrt(math.pow(math.max(qx, 0), 2) + math.pow(math.max(qy, 0), 2));
    return outside + math.min(math.max(qx, qy), 0) - r <= 0;
  }

  var lo = 0.0;
  var hi = math.sqrt(hw * hw + hh * hh);
  for (var i = 0; i < 32; i++) {
    final mid = (lo + hi) / 2;
    if (within(mid)) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  return lo;
}

class DialTick {
  const DialTick(this.x0, this.y0, this.x1, this.y1, this.width);
  final double x0;
  final double y0;
  final double x1;
  final double y1;
  final double width;
}

class DialNumeral {
  const DialNumeral(this.label, this.x, this.y);
  final String label;

  /// Centre of the numeral (its middle baseline, as the web draws it).
  final double x;
  final double y;
}

/// The dial's marks and numeral centres for a box of w × h, `s` units per mm
/// (px or mm alike). [measure] gives a label's width at font size `px`.
({List<DialTick> ticks, List<DialNumeral> numerals}) layoutDial(
  DialSource dial,
  double w,
  double h,
  double s,
  double Function(String label, double px) measure,
) {
  final d = math.min(w, h);
  final margin = d * 0.022;
  final hourTick = dial.ticks ? d * 0.055 : 0.0;
  final minuteTick = d * 0.026;
  final corner = dial.cornerRadiusMm * s;
  final ticks = <DialTick>[];
  if (dial.ticks) {
    for (var i = 0; i < 60; i++) {
      final angle = i * math.pi / 30;
      final hour = i % 5 == 0;
      final edge = dialEdgeDistance(dial.face, w, h, angle, corner) - margin;
      final len = hour ? hourTick : minuteTick;
      final dx = math.sin(angle);
      final dy = -math.cos(angle);
      ticks.add(DialTick(dx * edge, dy * edge, dx * (edge - len), dy * (edge - len), hour ? d * 0.011 : d * 0.0045));
    }
  }
  final numerals = <DialNumeral>[];
  if (dial.numerals != 'none') {
    final px = dial.sizeMm * s;
    for (var i = 0; i < 12; i++) {
      final angle = i * math.pi / 6;
      final label = dialLabel(dial, i);
      final width = measure(label, px);
      final dx = math.sin(angle);
      final dy = -math.cos(angle);
      final half = dx.abs() * (width / 2) + dy.abs() * (px * 0.42);
      final r = dialEdgeDistance(dial.face, w, h, angle, corner) - margin - hourTick - d * 0.02 - half;
      numerals.add(DialNumeral(label, dx * r, dy * r + px * 0.04));
    }
  }
  return (ticks: ticks, numerals: numerals);
}

/// Whether a press (in the layer's own frame, mm) lands on a numeral or a mark.
bool dialHit(
  DialSource dial,
  double w,
  double h,
  double px,
  double py,
  double Function(String label, double px) measure, {
  double slack = 2,
}) {
  final d = math.min(w, h);
  final margin = d * 0.022;
  final hourTick = dial.ticks ? d * 0.055 : 0.0;
  final minuteTick = d * 0.026;
  final corner = dial.cornerRadiusMm;
  final r = math.sqrt(px * px + py * py);
  if (dial.ticks) {
    final angle = math.atan2(px, -py);
    final minute = (angle * 30 / math.pi).round();
    final ray = minute * math.pi / 30;
    final across = (r * math.sin(angle - ray)).abs();
    final edge = dialEdgeDistance(dial.face, w, h, ray, corner) - margin;
    final len = minute % 5 == 0 ? hourTick : minuteTick;
    if (across <= slack && r <= edge + slack && r >= edge - len - slack) return true;
  }
  if (dial.numerals != 'none') {
    final size = dial.sizeMm;
    for (var i = 0; i < 12; i++) {
      final angle = i * math.pi / 6;
      final label = dialLabel(dial, i);
      final width = measure(label, size);
      final dx = math.sin(angle);
      final dy = -math.cos(angle);
      final half = dx.abs() * (width / 2) + dy.abs() * (size * 0.42);
      final at = dialEdgeDistance(dial.face, w, h, angle, corner) - margin - hourTick - d * 0.02 - half;
      if ((px - dx * at).abs() <= width / 2 + slack && (py - (dy * at + size * 0.04)).abs() <= size * 0.6 + slack) {
        return true;
      }
    }
  }
  return false;
}

/// The face an area's numerals follow.
({String face, double corner}) faceOf(PrintArea area) =>
    (face: area.round ? 'round' : 'rect', corner: area.cornerRadiusMm);

/// A new clock face's numerals, covering the colour-print zone, locked by the Studio.
Layer? newDialLayer(PrintArea area, String color) {
  final m = area.method('uv') ?? (area.methods.isEmpty ? null : area.methods.first);
  if (m == null) return null;
  final z = m.zone;
  final f = faceOf(area);
  return Layer(
    id: '$dialPrefix${DateTime.now().millisecondsSinceEpoch.toRadixString(36)}',
    area: area.key,
    method: m.method,
    kind: LayerKind.dial,
    locked: LayerLock.system,
    x: z.cx,
    y: z.cy,
    w: z.width,
    h: z.height,
    dial: DialSource(
      font: 'Montserrat',
      sizeMm: math.max(6, (math.min(z.width, z.height) * 0.075).roundToDouble()),
      color: m.colorsAllowed ? color : monoColor,
      bold: true,
      face: f.face,
      cornerRadiusMm: f.corner,
    ),
  );
}

/// Clock numerals follow their face on any type.
DesignDocument refitDials(DesignDocument doc, List<PrintArea> areas) => doc.copyWith(
      layers: [
        for (final l in doc.layers)
          if (l.dial != null && areaByKey(areas, l.area) != null)
            _refit(l, areaByKey(areas, l.area)!)
          else
            l,
      ],
    );

Layer _refit(Layer l, PrintArea area) {
  final fresh = newDialLayer(area, l.dial!.color);
  if (fresh == null) return l;
  final f = faceOf(area);
  return l.copyWith(
    x: fresh.x,
    y: fresh.y,
    w: fresh.w,
    h: fresh.h,
    rotation: 0,
    dial: l.dial!.copyWith(face: f.face, cornerRadiusMm: f.corner),
  );
}

/// A design without clock numerals gets them on every clock face.
DesignDocument withDials(DesignDocument doc, List<PrintArea> areas) {
  if (doc.layers.any((l) => l.dial != null)) return doc;
  final dials = [
    for (final a in areas)
      if (a.isDial) ?newDialLayer(a, contrastInk('#ffffff')),
  ];
  return dials.isEmpty ? doc : doc.copyWith(layers: [...dials, ...doc.layers]);
}
